athena-gcs/src/main/java/com/amazonaws/athena/connectors/gcs/storage/StorageMetadata.java

```diff
@@ -0,0 +1,315 @@
+import com.google.cloud.storage.Blob;
+import com.google.cloud.storage.Storage;
+import com.google.cloud.storage.StorageOptions;
+import com.google.common.collect.Lists;
+import org.apache.arrow.dataset.file.FileFormat;
+import org.apache.arrow.dataset.file.FileSystemDatasetFactory;
+import org.apache.arrow.dataset.jni.NativeMemoryPool;
+import org.apache.arrow.dataset.source.DatasetFactory;
+import org.apache.arrow.memory.BufferAllocator;
+import org.apache.arrow.vector.types.Types;
+import org.apache.arrow.vector.types.pojo.Field;
+import org.apache.arrow.vector.types.pojo.FieldType;
+import org.apache.arrow.vector.types.pojo.Schema;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.charset.StandardCharsets;
+import java.text.ParseException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.regex.Pattern;
+
+import static com.amazonaws.athena.connectors.gcs.GcsConstants.CLASSIFICATION_GLUE_TABLE_PARAM;
+import static com.amazonaws.athena.connectors.gcs.GcsConstants.PARTITION_PATTERN_KEY;
+import static com.amazonaws.athena.connectors.gcs.GcsUtil.createUri;
+import static com.amazonaws.athena.connectors.gcs.GcsUtil.isFieldTypeNull;
+import static com.google.cloud.storage.Storage.BlobListOption.prefix;
+import static java.util.Objects.requireNonNull;
+
+public class StorageMetadata
+{
+    private static final Logger LOGGER = LoggerFactory.getLogger(StorageMetadata.class);
+    private final Storage storage;
+
+    /**
+     * Instantiate a storage data source object with provided config
+     *
+     * @param gcsCredentialJsonString An instance of GcsDatasourceConfig that contains necessary properties for instantiating an appropriate data source
+     * @throws IOException If occurs during initializing input stream with GCS credential JSON
+     */
+    public StorageMetadata(String gcsCredentialJsonString) throws IOException
+    {
+        requireNonNull(gcsCredentialJsonString, "GCS credential JSON is null");
+        GoogleCredentials credentials
+                = GoogleCredentials.fromStream(new ByteArrayInputStream(gcsCredentialJsonString.getBytes(StandardCharsets.UTF_8)))
+                .createScoped(Lists.newArrayList("https://www.googleapis.com/auth/cloud-platform"));
+        storage = StorageOptions.newBuilder().setCredentials(credentials).build().getService();
+    }
+
+    /**
+     * Returns a list of field names and associated file type
+     *
+     * @param bucketName Name of the database
+     * @param tableName    Name of the table
+     * @param format   classification param form table
+     * @return An instance of {@link List<Field>} with column metadata
+     */
+    public synchronized List<Field> getFields(String bucketName, String tableName, String format, BufferAllocator allocator)
```

Comment:
Is there a reason that this method is `public synchronized` ?

It seems like the only thing calling this method is `buildTableSchema()` within this file right? If so this can just be made `private`.

Explain the need for `synchronized` if there's an actual need for it.

Reply:
not needed. so removed.